### PR TITLE
GH-95: [RTI][TOOL][BACKEND] RTI Request Management APIs

### LINKED ISSUE #56: [RTI][TOOL][BACKEND] Position List Retrieval API

### LINKED ISSUE #59: [RTI][TOOL][BACKEND] RTI Request List Retrieval API

### LINKED ISSUE #95: [RTI][TOOL][BACKEND] RTI Request Management APIs

### LINKED ISSUE #96: [RTI][TOOL][BACKEND] RTI Request Ready By ID API

### LINKED ISSUE #97: [RTI][TOOL][BACKEND] RTI Request Creation API

### LINKED ISSUE #98: [RTI][TOOL][BACKEND] RTI Request Deletion API

### LINKED ISSUE #99: [RTI][TOOL][BACKEND] RTI Request Update API

### PR DESCRIPTION
**Pull request by ChanukaUOJ**
## Overview
This PR introduces four new API endpoints that enables **rti_request** read, create, update, deletion and read by ID. 

**New API endpoint:**

`CREATE /rti_request`
`POST /rti_request`
`PUT /rti_request/{id}`
`GET /rti_request/{id}`
`DELETE /rti_request/{id}`

---

## Changes
* **Routing:** Added new routers to the `rti_request_router.py` file.
* **Service Layer:** Developed the services in the `rti_request_service.py` file.
* **Testing:** Added new test cases to cover the service functionalities.

---

## Testing
* **Service Tests:** Updated `test_rti_request_service.py` with new test cases to cover the new endpoint.
* **Mocks:** Updated `conftest.py` with new mock services.
* **Status:** All tests are passing.

---

## Additional Information
This PR Closes #95
Closes #56
Closes #96 
Closes #97
Closes #98 
Closes #99

### CODE REVIEW

File: tool/backend/src/models/table_schemas/table_schemas.py [hunk 1]
```diff
@@ -111,3 +119,80 @@ class Sender(SQLModel, table=True):
...
+
+    # table fields
+    id: UUID = Field(primary_key=True, description="Unique identifier for the RTI Status History")
+    rti_request_id: UUID = Field(foreign_key="rti_requests.id", description="Unique identifier for the RTI Request")
+    status_id: UUID = Field(foreign_key="rti_statuses.id", description="Unique identifier for the RTI Status")
+    direction: RTIDirection = Field(sa_column=Column(SAEnum(RTIDirection, name="rti_direction")), description="direction for the RTI Status History")
+    description: Optional[str] = Field(default=None, description="description for the RTI Status History")
+    entry_time: datetime = Field(description="entry time for the RTI Status History")
+    exit_time: Optional[datetime] = Field(default=None, description="exit time for the RTI Status History")
+    files: List[str] = Field(..., sa_column=Column(JSON), description="List of URLs for the RTI status history files")
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/models/table_schemas/table_schemas.py]: Noted!

File: tool/backend/src/models/table_schemas/table_schemas.py [hunk 2]
```diff
@@ -1,8 +1,13 @@
 from datetime import datetime, timezone
 from typing import Optional, List
 from uuid import UUID
+from enum import Enum
 from sqlmodel import SQLModel, Field, func, CheckConstraint, Relationship
+from sqlalchemy import Column, JSON, Enum as SAEnum
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/models/table_schemas/table_schemas.py]: Here we use JSON to keep it more generic so we can easily change the db driver later if we want

File: tool/backend/src/routers/rti_request_router.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    page_size: int = Query(10, ge=1, le=100, description="page size"),
+    service: RTIRequestService = Depends(get_rti_request_service),
+    user: User = Depends(RoleChecker([UserRole.ADMIN, UserRole.USER]))
+):
+    response = service.get_rti_requests(page=page, page_size=page_size)
+    return response
+
+@router.get("/rti_requests/{id}", response_model=RTIRequestResponse)
+def get_rti_request_by_id_endpoint(
+    id: Annotated[str, Path(title="ID of the RTI Request")],
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/routers/rti_request_router.py]: Noted!

File: tool/backend/src/services/rti_request_service.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            )
+        except Exception as e:
+            logger.error(f"Error fetching RTI requests: {e}")
+            raise InternalServerException("Failed to fetch RTI requests from database.") from e
+
+    # API
+    def get_rti_request_by_id(
+        self,
+        *,
+        request_id
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/src/routers/rti_request_router.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    request_data = RTIRequestRequest(
+        title=title,
+        sender_id=sender_id,
+        receiver_id=receiver_id,
+        file=file,
+        description=description,
+        rti_template_id=rti_template_id
+    )
+    response = await service.create_rti_request(request_data=request_data)
+    return response
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/routers/rti_request_router.py:58]: Added!

File: tool/backend/src/services/rti_request_service.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            unique_id = uuid4()
+            uploaded_file_path: str | None = None
+
+            # 1. validate file extension
+            if not request_data.file:
+                raise BadRequestException("RTI Request file is required")
+
+            _, ext = os.path.splitext(request_data.file.filename)
+            if not ext or ext.lower() not in [".pdf"]:
+                raise BadRequestException(f"{request_data.file.filename} doesn't have a valid extension (.pdf)")
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/src/services/rti_request_service.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        *,
+        page: int = 1,
+        page_size: int = 10
+    ) -> RTIRequestListResponse:
+        """Fetches a paginated list of RTI Requests."""
+        try:
+            offset = (page - 1) * page_size
+
+            # fetch the records from the table
+            statement_records = select(RTIRequest).offset(offset).limit(page_size)
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/src/services/rti_request_service.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+            unique_id = uuid4()
+            uploaded_file_path: str | None = None
+
+            # 1. validate file extension
+            if not request_data.file:
+                raise BadRequestException("RTI Request file is required")
+
+            _, ext = os.path.splitext(request_data.file.filename)
+            if not ext or ext.lower() not in [".pdf"]:
+                raise BadRequestException(f"{request_data.file.filename} doesn't have a valid extension (.pdf)")
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/src/models/table_schemas/table_schemas.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/models/table_schemas/table_schemas.py]: Noted!

File: tool/backend/src/services/rti_request_service.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+                description=request_data.description,
+                sender_id=request_data.sender_id,
+                receiver_id=request_data.receiver_id,
+                rti_template_id=request_data.rti_template_id
+            )
+            self.session.add(rti_request)
+
+            # 4. Insert RTIStatusHistories
+            statement = select(RTIStatus).where(RTIStatus.name == "CREATED")
+            created_status = self.session.exec(statement).first()
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/src/services/rti_request_service.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+                if history.files:
+                    all_file_paths.extend(history.files)
+
+            old_files_backup = []
+
+            # 2. Backup and delete files from GitHub
+            try:
+                for file_path in all_file_paths:
+                    file_data = await self.file_service.read_file(file_path)
+                    old_files_backup.append({"path": file_path, "content": file_data["content"]})
```
[the remark this answers is not in the document]
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/sql/seed.sql [hunk 11]
```diff
@@ -90,34 +92,34 @@ INSERT INTO rti_requests (title, description, sender_id, receiver_id, rti_templa
...
+((SELECT id FROM rti_requests WHERE title = 'School Expenditure 2024' LIMIT 1), (SELECT id FROM rti_statuses WHERE name = 'SENT_FOR_APPROVAL' LIMIT 1), 'sent', 'Resubmitted with correct documents.', NOW() - INTERVAL '18 days', NOW() - INTERVAL '17 days', '["requests/req_002_v2.pdf"]'),
+((SELECT id FROM rti_requests WHERE title = 'School Expenditure 2024' LIMIT 1), (SELECT id FROM rti_statuses WHERE name = 'APPROVED' LIMIT 1), 'sent', 'Verified and approved.', NOW() - INTERVAL '17 days', NOW() - INTERVAL '16 days', '[]'),
+((SELECT id FROM rti_requests WHERE title = 'School Expenditure 2024' LIMIT 1), (SELECT id FROM rti_statuses WHERE name = 'SENT_TO_RECEIVER' LIMIT 1), 'sent', 'Sent to Dept of Education.', NOW() - INTERVAL '16 days', NOW() - INTERVAL '5 days', '[]'),
+((SELECT id FROM rti_requests WHERE title = 'School Expenditure 2024' LIMIT 1), (SELECT id FROM rti_statuses WHERE name = 'COMPLETED' LIMIT 1), 'received', 'Final response received and shared.', NOW() - INTERVAL '5 days', NOW() - INTERVAL '5 days', '["responses/data_002.zip"]');
```
Comment (zaeema-n) [inline tool/backend/sql/seed.sql:107]: This fails because the names don't match the names in the rti_statuses table

File: tool/backend/src/services/rti_request_service.py [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+                    RTIStatusHistories.rti_request_id == target_id,
+                    RTIStatusHistories.status_id == created_status.id
+                )
+                status_history = self.session.exec(history_statement).first()
+
+                if not status_history or not status_history.files:
+                    raise InternalServerException("Initial file record not found for this RTI Request.")
+
+                old_file_path = status_history.files[0]
+                content = await request_data.file.read()
```
Comment (zaeema-n) [inline tool/backend/src/services/rti_request_service.py]:
```suggestion
                new_file_content = await request_data.file.read()
```

File: tool/backend/src/services/rti_request_service.py [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+        try:
+            unique_id = uuid4()
+            uploaded_file_path: str | None = None
+
+            # 1. validate file extension
+            if not request_data.file:
+                raise BadRequestException("RTI Request file is required")
+
+            _, ext = os.path.splitext(request_data.file.filename)
+            if not ext or ext.lower() not in [".pdf"]:
```
Comment (zaeema-n) [inline tool/backend/src/services/rti_request_service.py]: We should keep the allowed file types for rti requests in a constant, then we can easily add new file types.

Comment (zaeema-n) [reply tool/backend/src/services/rti_request_service.py]: Same can be done for rti templates and any others that use files

File: tool/backend/src/services/rti_request_service.py [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+                    await self.file_service.delete_file(file_path=new_file_path)
+                except Exception as ex:
+                    logger.error(f"[RTI SERVICE] Rollback failed (delete new): {ex}")
+
+            if isinstance(e, IntegrityError):
+                logger.error(f"[RTI SERVICE] Integrity error updating RTI request: {e}")
+                raise ConflictException("Database integrity error occurred while updating the RTI Request.") from e
+
+            logger.error(f"[RTI SERVICE] Error updating RTI request: {e}")
+            raise InternalServerException(f"Failed to update RTI request: {e}") from e
```
Comment (zaeema-n) [inline tool/backend/src/services/rti_request_service.py:321]: There is a consistency bug risk in `update_rti_request` because rollback compensation runs even when failure happens after DB commit.

Failure sequence:

1. Request updates file with a different extension (`new_file_path` is created; DB history is updated to point to new path).
2. `self.session.commit()` succeeds, so DB now permanently references new_file_path.
3. Next step tries to delete old file (`delete_file(old_file_path)`) and this call fails (network/GitHub/API issue).
4. Control enters the broad `except Exception` block.
5. Rollback logic executes `elif new_file_path: delete_file(new_file_path)`.
6. Result: DB still points to `new_file_path`, but that file has now been deleted by compensation.
7. System ends in inconsistent state (broken file reference/data loss from app perspective).

Suggested fix: track commit state (e.g. `committed = True` after successful commit) and only run compensating file rollback for pre-commit failures. Post-commit cleanup failures should be handled separately (log/retry/orphan cleanup), not by deleting the newly committed file.

File: tool/backend/src/services/rti_request_service.py [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+                raise NotFoundException(f"RTI Request with id {request_id} not found.")
+
+            # 1. Fetch all histories and their files
+            statement = select(RTIStatusHistories).where(RTIStatusHistories.rti_request_id == target_id)
+            histories = self.session.exec(statement).all()
+            
+            # If there are more than 1 history record, it means the request has progressed
+            # and should not be deleted according to business rules.
+            if len(histories) > 1:
+                raise ConflictException("Cannot delete RTI Request because it has associated status history records beyond creation.")
```
Comment (zaeema-n) [inline tool/backend/src/services/rti_request_service.py:343]: If the user wants to delete a request can't we assume they also want to delete the associated status histories? For example, they sent the request for approval and was then told they no longer need it.

File: tool/backend/test/conftest.py [hunk 16]
```diff
@@ -473,3 +473,135 @@ def _factory(name: str = "Test Position") -> MagicMock:
...
+        if description: fields["description"] = description
+        if sender_id: fields["sender_id"] = sender_id
+        if receiver_id: fields["receiver_id"] = receiver_id
+        if rti_template_id: fields["rti_template_id"] = rti_template_id
+        
+        # Re-initialize with only the provided fields to ensure exclude_unset works
+        request = RTIRequestUpdateRequest(**fields)
+        if not id:
+            request.id = str(uuid.uuid4()) # Ensure ID is always set for service fetch
+            request.__pydantic_fields_set__.add("id")
```
Comment (zaeema-n) [inline tool/backend/test/conftest.py]: Why do we do this?

File: tool/backend/test/test_rti_request_service.py [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+from unittest.mock import MagicMock, AsyncMock, patch
+from sqlalchemy.exc import IntegrityError, OperationalError
+from sqlalchemy import event
+from sqlmodel import SQLModel, Session, create_engine, select
+from datetime import datetime, timezone
+
+from src.services.rti_request_service import RTIRequestService
+from src.models.table_schemas.table_schemas import (
+    RTIRequest, RTIStatus, RTIStatusHistories, RTIDirection, 
+    Sender, Receiver, Position, Institution
```
Comment (zaeema-n) [inline tool/backend/test/test_rti_request_service.py]: Position, Institution not used

File: tool/backend/test/test_rti_request_service.py [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+)
+from src.models.request_models.rti_requests import RTIRequestRequest, RTIRequestUpdateRequest
```
Comment (zaeema-n) [inline tool/backend/test/test_rti_request_service.py]: RTIRequestRequest, RTIRequestUpdateRequest not used

File: tool/backend/test/test_rti_request_service.py [hunk 19]
```diff
@@ -0,0 +1,438 @@
+# tests/test_rti_request_service.py
+import uuid
+import pytest
+import re
+from unittest.mock import MagicMock, AsyncMock, patch
+from sqlalchemy.exc import IntegrityError, OperationalError
+from sqlalchemy import event
```
Comment (zaeema-n) [inline tool/backend/test/test_rti_request_service.py]: event not used

File: tool/backend/test/test_rti_request_service.py
Comment (zaeema-n) [inline tool/backend/test/test_rti_request_service.py:1]: - Missing update_rti_request test for post-commit failure (e.g., old-file delete fails after DB commit) that can cause DB/file inconsistency.
- Missing update_rti_request test for blocked update when history count > 1.
- Missing update validation tests: invalid UUID, invalid sender_id/receiver_id/rti_template_id.
- Missing update file-edge tests: invalid extension, missing CREATED status, missing initial history file.
- Missing create edge tests: file missing, invalid template id, file service returns empty relative_path, IntegrityError -> ConflictException.
- Missing delete edge tests: invalid UUID and file cleanup failure after successful DB delete should not fail request.

File: tool/backend/sql/seed.sql (hunk 11 above)
Comment (ChanukaUOJ) [reply tool/backend/sql/seed.sql:107]: Noted!

File: tool/backend/src/services/rti_request_service.py (hunk 12 above)
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted!

File: tool/backend/test/test_rti_request_service.py (hunk 17 above)
Comment (ChanukaUOJ) [reply tool/backend/test/test_rti_request_service.py]: Noted!

File: tool/backend/test/test_rti_request_service.py (hunk 18 above)
Comment (ChanukaUOJ) [reply tool/backend/test/test_rti_request_service.py]: Noted!

File: tool/backend/test/test_rti_request_service.py (hunk 19 above)
Comment (ChanukaUOJ) [reply tool/backend/test/test_rti_request_service.py]: Noted!

File: tool/backend/src/services/rti_request_service.py (hunk 15 above)
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py:343]: My opinion is If it is a newly created one we can delete it including the CREATED history. but if it is has more statuses we should not delete it to show the progress of that. if we want to handle this type of cases we can add that to a "DISABLED" state. @zaeema-n

File: tool/backend/test/conftest.py (hunk 16 above)
Comment (ChanukaUOJ) [reply tool/backend/test/conftest.py]: This is to ensure that "id" field is always there in the request in the initialization. but that is not something required according to services because in through model dump we don't set or use the id. bcz of that this will be removed!

File: tool/backend/src/services/rti_request_service.py (hunk 13 above)
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: Noted! in this PR this will be covered only for rti request service.

Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py]: https://github.com/LDFLK/RTI-Tracker/issues/106

created this issue to cover this in other services

File: tool/backend/test/test_rti_request_service.py
Comment (ChanukaUOJ) [reply tool/backend/test/test_rti_request_service.py:1]: Noted!

File: tool/backend/src/services/rti_request_service.py (hunk 14 above)
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py:321]: Noted!

File: tool/backend/src/services/rti_request_service.py (hunk 15 above)
Comment (zaeema-n) [reply tool/backend/src/services/rti_request_service.py:343]: So the DISABLED state will go into the status_histories table?

**Review by zaeema-n — APPROVED** [review]
LGTM

File: tool/backend/src/services/rti_request_service.py (hunk 15 above)
Comment (ChanukaUOJ) [reply tool/backend/src/services/rti_request_service.py:343]: yes!
